pkg/lib/bundlev2/generate.go

```diff
@@ -0,0 +1,144 @@
+package bundlev2
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"gopkg.in/yaml.v2"
+	"k8s.io/kubectl/pkg/util/slice"
+)
+
+const SynthesizedCSVFileName string = "olm.clusterserviceversion.yaml"
+
+// GenerateFunc generates a CSV for the bundle based on the resource within the bundle directory.
+// The bundleDir points to where the bundle is located, outputDir is where bundle and synthesized CSV is copy/write to.
+// If overwrite is set to true, outputDir is reloaded with the exact content of bundleDir plus a synthesized CSV
+// and inplace generation deletes the CSV if exists.
+func GenerateFunc(bundleDir, outputDir string, overwrite bool) error {
+	bundleDir, err := filepath.Abs(bundleDir)
+	if err != nil {
+		return err
+	}
+
+	if outputDir != "" {
+		outputDir, err = filepath.Abs(outputDir)
+		if err != nil {
+			return err
+		}
+	} else {
+		// if outputDir not specified, synthesis within the bundle directory.
+		outputDir = bundleDir
+	}
+
+	bundle, err := NewBundleV2FromDirectory(bundleDir)
+	if err != nil {
+		return err
+	}
+
+	csv, err := bundle.SynthesisedCSV()
+	if err != nil {
+		return err
+	}
+
+	if bundle.existingCSVPath != "" && !overwrite {
+		return fmt.Errorf("a ClusterServiceVersion already exist in the bundle at %s", bundle.existingCSVPath)
+	}
+
+	// Only delete (overwrite) the existing CSV in a bundle if specified to overwrite for inplace generation.
+	if bundle.existingCSVPath != "" && outputDir == bundleDir {
+		if err := os.Remove(bundle.existingCSVPath); err != nil {
+			return fmt.Errorf("error overwriting ClusterServiceVersion %s, %v", bundle.existingCSVPath, err)
+		}
+	} else if outputDir != bundleDir {
+		if overwrite {
+			// Do not expect path error since path is examined earlier.
+			os.RemoveAll(outputDir)
+		}
+		if err := CopyAllFilsFromDirectoryExcept(bundleDir, outputDir, []string{bundle.existingCSVPath}); err != nil {
+			return err
+		}
+	}
+
+	// generate CSV file to the outputDir.
+	return WriteStructToFile(csv, SynthesizedCSVFileName, outputDir)
+}
+
+// WriteStructToFile write a struct content to YAML a file.
```

Comment:
```suggestion
// WriteStructToFile write a struct's content to a YAML file.
```

But this is typically called `Marshal` as well.